### PR TITLE
test: aws accounts integ test

### PR DESCRIPTION
**Pull request by SanketD92**
Issue #, if available:
GALI-1575

Description of changes:
Integration test for AWS Account onboarding.

Checklist:

* [x] Have you successfully deployed to an AWS account with your changes?
* [x] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: solutions/swb-reference/integration-tests/config/example.yaml [hunk 1]
```diff
@@ -2,15 +2,22 @@
...
 #terminatedEnvId: ''   # id of an environment that has already been terminated
 
-#Cognito Integ Test Client
+## Cognito Integ Test Client
 #rootUsername: ''
 #rootPasswordParamStorePath: ''
+
+## Pre-requisite for running aws-accounts integration test
+## Create the onboard_account stack in an external AWS account, and populate the following values below:
+# awsAccountId: ''
```
Comment (nguyen102) [inline solutions/swb-reference/integration-tests/config/example.yaml:20]: Nitpick: Delete the extra space between `#` and `a`. Same for the lines 21 to 23

File: solutions/swb-reference/integration-tests/support/complex/accountHelper.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  private _settings: Settings;
+  public constructor(awsSdkClient: AwsService) {
+    const setup: Setup = new Setup();
+    const settings: Settings = setup.getSettings();
+    this._settings = settings;
+    this._awsSdk = awsSdkClient;
+  }
+
+  public async listOnboardedAccounts(): Promise<{[id: string]: string}[]> {
+    const queryParams = {
```
Comment (nguyen102) [inline solutions/swb-reference/integration-tests/support/complex/accountHelper.ts:20]: This should probably just be an API call to list accounts. If we don't have that API yet, can we add a TODO to this code segment to replace this with just calling the API.

File: solutions/swb-reference/integration-tests/support/complex/accountHelper.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  public async deleteDdbRecords(accountId: string, awsAccountId: string): Promise<void> {
+    await this._awsSdk.helpers.ddb
+      .delete({ pk: `ACC#${accountId}`, sk: `ACC#${accountId}` }).execute();
+    await this._awsSdk.helpers.ddb
+      .delete({ pk: `AWSACC#${awsAccountId}`, sk: `ACC#${accountId}` }).execute();
+  }
+
+  public async verifyBusPermitsAccount(awsAccountId: string): Promise<boolean> {
```
Comment (nguyen102) [inline solutions/swb-reference/integration-tests/support/complex/accountHelper.ts:41]: Should `permits` be `permission`?

Typo with the `t`?

File: solutions/swb-reference/integration-tests/support/complex/accountHelper.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    // Update bucket policy
+    await this._awsSdk.clients.s3.putBucketPolicy(putPolicyParams);
+  }
+
+  public async deOnboardAccount(awsAccountId: string): Promise<void> {
+  // Undo all operations that happen in: hostingAccountLifecycleService.initializeAccount()
+
+  // Update main account default event bus to remove hosting account state change events
+  await this.updateBusPermissions(awsAccountId);
```
Comment (nguyen102) [inline solutions/swb-reference/integration-tests/support/complex/accountHelper.ts:145]: Should we change the `updateBusPermission` name to describe more specifically what it does? The comment on line 144 is very helpful. Devs shouldn't need to look at the `deOnboardAccount` method to understand what the `updateBusPermission` method does.

File: solutions/swb-reference/integration-tests/support/complex/accountHelper.ts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+  // Remove account to artifactBucket's bucket policy
+  await this.updateArtifactsBucketPolicy(awsAccountId);
```
Comment (nguyen102) [inline solutions/swb-reference/integration-tests/support/complex/accountHelper.ts:148]: Let's incorporate part of the comment in line 147 into `updateArtifactsBucketPolicy` name. Maybe `removeAccountFromArtifactBucketPolicy`

File: solutions/swb-reference/integration-tests/support/complex/accountHelper.ts (hunk 3 above)
Comment (SanketD92) [reply solutions/swb-reference/integration-tests/support/complex/accountHelper.ts:41]: It's not a typo, "permits" is used as a verb. This method verifies if the event bus _permits_ the hosting account to access it.

File: solutions/swb-reference/integration-tests/tests/multiStep/aws-account.test.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
+/*
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+import _ from 'lodash';
+// import { v4 as uuidv4 } from 'uuid';
```
[the remark this answers is not in the document]
Comment (SanketD92) [reply solutions/swb-reference/integration-tests/tests/multiStep/aws-account.test.ts:6]: Oops

File: solutions/swb-reference/integration-tests/tests/multiStep/aws-account.test.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      }
+      return currConfig;
+    });
+  }
+
+  test('Onboarding new hosting account', async () => {
+    if (!canTestAccountsApi()) {
+      console.log('Config settings needed to test AWS Account onboarding are missing. Skipping this test')
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (SanketD92) [reply solutions/swb-reference/integration-tests/tests/multiStep/aws-account.test.ts:48]: We are using `settings.get()` in the `canTestAccountsApi()` method but we're checking for all the required settings. 

This integration test is a bit more involved in that it requires a user to manually create a CloudFormation stack in a new hosting account and provide its output values. I'm not failing the test if the user does not do that, rather it passes with a warning that configs weren't provided so it has been skipped.

File: solutions/swb-reference/integration-tests/support/utils/settings.ts [hunk 8]
```diff
@@ -72,8 +78,9 @@ export default class Settings {
     return value;
   }
 
-  public optional(key: SettingKey, defaultValue: string): string {
+  public optional(key: SettingKey, defaultValue?: string): string | undefined {
     const value = this._content[key];
+    console.log(value);
```
Comment (nguyen102) [inline solutions/swb-reference/integration-tests/support/utils/settings.ts:83]: Delete?

**Review by nguyen102 — APPROVED** [review]

**Review by insignias — APPROVED** [review]

**Pull request merged**